src/Npgsql/NpgsqlTypes/NpgsqlDate.cs

```diff
@@ -108,72 +108,204 @@ public NpgsqlDate(int year, int month, int day)
+        /// </exception>
+        /// <remarks>
+        /// See: https://www.postgresql.org/docs/current/static/datatype-datetime.html
+        /// </remarks>
+        public static NpgsqlDate Parse([NotNull] string input)
+        {
+            if (input is null)
+                throw new ArgumentNullException(nameof(input));
+
+            var str = input.AsSpan().Trim();
+
+            if (str.SequenceEqual(InfinityLiteral.AsSpan()))
                 return Infinity;
 
-            if (str == "-infinity")
+            if (str.SequenceEqual(NegativeInfinityLiteral.AsSpan()))
                 return NegativeInfinity;
 
-            str = str.Trim();
-            try {
-                var idx = str.IndexOf('-');
-                if (idx == -1) {
-                    throw new FormatException();
-                }
-                var year = int.Parse(str.Substring(0, idx));
-                var idxLast = idx + 1;
-                if ((idx = str.IndexOf('-', idxLast)) == -1) {
-                    throw new FormatException();
-                }
-                var month = int.Parse(str.Substring(idxLast, idx - idxLast));
-                idxLast = idx + 1;
-                if ((idx = str.IndexOf(' ', idxLast)) == -1) {
-                    idx = str.Length;
-                }
-                var day = int.Parse(str.Substring(idxLast, idx - idxLast));
-                if (str.Contains("BC")) {
+            try
+            {
+                var yearSeparator = str.IndexOf('-');
+
+                if (yearSeparator == -1)
+                    throw new FormatException($"Malformed date literal: {str.ToString()}");
+
+                var monthSeparator = str.Slice(yearSeparator + 1).IndexOf('-');
+
+                if (monthSeparator == -1)
+                    throw new FormatException($"Malformed date literal: {str.ToString()}");
+
+                var daySeparator = str.Slice(yearSeparator + monthSeparator + 1).IndexOf(' ');
+
+                var yearSegment = str.Slice(0, yearSeparator);
+                var monthSegment = str.Slice(yearSeparator + 1, monthSeparator);
+                var daySegment =
+                    daySeparator == -1
+                        ? str.Slice(yearSeparator + monthSeparator + 2)
+                        : str.Slice(yearSeparator + monthSeparator + 2, daySeparator);
+
+                var year = int.Parse(yearSegment.ToString());
+                var month = int.Parse(monthSegment.ToString());
+                var day = int.Parse(daySegment.ToString());
+
+                if (str.Contains("BC".AsSpan(), StringComparison.Ordinal))
                     year = -year;
-                }
+
                 return new NpgsqlDate(year, month, day);
-            } catch (OverflowException) {
+            }
+            catch (OverflowException)
+            {
                 throw;
-            } catch (Exception) {
-                throw new FormatException();
+            }
+            catch (Exception)
```

Comment:
I hate how exception handling is being done here (and also that `TryParse()` calls `Parse()` and catches any exceptions to return `false`). Any chance you want to make this better?

IMHO the parsing code should be in `TryParse()`, which itself calls only the `int.TryParse()`. Then `Parse()` wraps `TryParse()` and throws a FormatException if it gets a false.

Reply:
Very much agree on the exception handling. I'll see what I can do WRT reorganizing `TryParse`/`Parse`.